### PR TITLE
refactor(compiler-cli): Track external component resources in Resourc…

### PR DESCRIPTION
**Pull request by atscott**
…eRegistry

In addition to the template mapping that already existed, we want to also track the mapping for external
style files. We also store the `ts.Expression` in the registry so external tools can look up a resource
on a component by expression and avoid reading the value.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 1]
```diff
@@ -641,20 +638,38 @@ export class ComponentDecoratorHandler implements
...
-          styleUrlsExpr, styleUrls, 'styleUrls must be an array of strings');
+    // If styleUrls is a literal array of strings, process each resource url individually and
+    // register them. Otherwise, simply evaluate the entire expression at once with the
+    // `PartialEvaluator`.
+    if (ts.isArrayLiteralExpression(styleUrlsExpr) &&
+        styleUrlsExpr.elements.every(e => ts.isStringLiteralLike(e))) {
+      const styleUrls: string[] = [];
+      const containingFile = componentClass.getSourceFile().fileName;
+      for (const expression of styleUrlsExpr.elements) {
+        const resolved = this.evaluator.evaluate(expression) as string;
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts:657]: I don't think it's necessary to use the evaluator here - if you've asserted that the `ts.Expression` is `ts.isStringLiteralLike`, you should be able to grab the text of the literal directly.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 2]
```diff
@@ -385,10 +386,6 @@ export class ComponentDecoratorHandler implements
       ...analysis.typeCheckMeta,
     });
 
-    if (!analysis.template.isInline) {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts:388]: There's a problem with the way this is refactored, related to incremental builds.

`analyze` is an operation that's performed incrementally. When a `ts.SourceFile` has changed, `analyze` is run against all of the decorated classes within it, and it returns the metadata extracted from that single class. If the `ts.SourceFile` hasn't changed, `analyze` does not run, and the metadata from the previous program's analysis is reused.

This means that `analyze` should not be side-effectful - we can't rely on it to register information about the component/directive in any compiler data structures. The `register` operation of a `DecoratorHandler` exists for this reason: `register` _always_ runs for all classes, regardless of whether or not they've changed in the current incremental step. It should therefore be as efficient as possible, and only take information in the analysis metadata and register it with the rest of the compiler.

Therefore the `ResourceRegistry` population needs to happen in two steps:
1) during `analyze`, information about resources needs to be extracted from the decorator and stored in the analysis metadata
2) during `register`, that information can then be submitted to the `ResourceRegistry`

This ensures that during each incremental step, the `ResourceRegistry` will always have resources for every component, regardless of whether or not the component has changed.

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts [hunk 3]
```diff
@@ -239,8 +239,27 @@ export class NgCompiler {
...
+
+  /**
+   * Retrieves the `ts.Declaration`s for any component(s) which use the given template file.
+   */
+  getComponentsWithStyleFile(styleFilePath: string): ReadonlySet<DeclarationNode> {
+    const {resourceRegistry} = this.ensureAnalyzed();
+    return resourceRegistry.getComponentsWithStyle(resolve(styleFilePath));
+  }
+
+  getComponentResources(classDecl: DeclarationNode) {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/core/src/compiler.ts:254]: Nit: explicit return types

**Review by alxhub — CHANGES_REQUESTED** [review]
Awesome work! One blocking comment, but it shouldn't be too tricky to resolve.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts (hunk 2 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/annotations/src/component.ts:388]: Done. PTAL

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts (hunk 1 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: I don't know if this indicates there's another problem, but when I removed the `evaluate` and just used `expression.getText()`, several tests failed, such as the `ngtsc behavioral tests should compile components with styleUrls`.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 4]
```diff
@@ -237,6 +239,7 @@ export class ComponentDecoratorHandler implements
     // Extract a closure of the template parsing code so that it can be reparsed with different
     // options if needed, like in the indexing pipeline.
     let template: ParsedTemplateWithSource;
+    let templateResource: Resource|null = null;
     if (this.preanalyzeTemplateCache.has(node)) {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts:243]: You probably also need to handle this case, where the template is parsed asynchronously (in `preanalyze`). Alternatively, extract the information you need from the `template` structure following this whole `if` block.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 5]
```diff
@@ -287,6 +291,7 @@ export class ComponentDecoratorHandler implements
     // component.
     let styles: string[]|null = null;
 
+    const styleResources = this._extractStyleResources(component, containingFile);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts:294]: This looks like it's duplicating some of the work of `_extractStyleUrls`. We should combine the two operations.

**Review by alxhub — APPROVED** [review]

**Pull request closed without merging**